Script execution order

RichardKarlFuchs opened the pull request.

RichardKarlFuchs committed:
add update publisher

funketh committed:
Merge branch 'develop' into feature/script_execution_order

funketh committed:
don't try to call the update events if they are null

- prevents NullReferenceException

funketh committed:
remove unused import

funketh committed:
shorten code using expression bodies

funketh committed:
use explicit private modifier

funketh committed:
move UpdatePublisher.cs into its own Events/ namespace

funketh commented:
> Ich glaube nicht, dass das funktioniert, weil unity ja auf die Funktion zugreifen muss. Bin aber dessen nicht sicher

Wenn man einen access modifier weglässt, wählt C# den möglichst restriktiven modifier, also in dem Fall sowieso private (habe es nur explizit gemacht).

Außerdem ruft unity die nicht direkt auf, sondern wahrscheinlich irgendeine methode in MonoBehaviour, die das dann intern löst. Wahrscheinlich.
Nachgucken kann ichs natürlich nicht, weil Unity proprietary garbage ist.

rrickfox commented:
Kannst es ja mal checken, indem du ein `Debug.Log()` mit irgendwas drin aufrufst. Und ein GaneObject mit dem Script drauf erstellst, falls es noch nicht existiert.

funketh commented:
Nein, ich bin mir schon sicher, dass es funktioniert. Wie gesagt, wenn mans weglässt ist es eh automatisch private. So funktioniert C# nunmal.

Ich bezog mich nur darauf, dass ich nicht nachschauen kann wie die Update Methoden tatsächlich gecalled werden (was jetzt nicht direkt für dieses Projekt relevant ist).

funketh committed:
remove LinkedBehaviour and use subscription pattern in DataTypes

funketh committed:
use experimental weak references (does not work yet)

funketh commented on Assets/Scripts/Events/WeakVoidPublisher.cs:
```diff
@@ -0,0 +1,37 @@
+                                    " other subscribers to the same WeakVoidPublisher", e);
+            }
+        }
+    }
```
Problem: Autos bleiben sofort für immer stehen. Schwer im Debugger reproduzierbar, da sich die Autos beim Debuggen erst etwas bewegen, bevor sie auch für immer stehen bleiben.

RichardKarlFuchs replied:
Welchen Vorteil hat denn das WeakRefefencing gegenüber der vorherigen Methode?

funketh replied:
Zeile 6:
```cs
// Publisher that does not keep subscribers alive if they are only referenced by this publisher
```
Die Autos werden sonst von den Referenzen im Publisher für immer am Leben erhalten, und der Publisher ruft dann z.B. Move() bei allen Autos auf, die eigentlich schon despawned sind und deren GameObject schon gelöscht (null) ist.

Die schwachen Referenzen halten die Auto Objekte nicht am Leben und wenn ein Auto zerstört wird, wird das Target von der WeakReference einfach zu null.

RichardKarlFuchs replied:
Wenn die Autos gelöscht werden, wird ja ihre `Dispose()` Funktion aufgerufen, die jedes GameObjectData hat. Dann kann man sich doch in dieser Funktion von allen Events unsubscriben.

funketh replied:
Das wäre die alternative Lösung. Ist halt sehr hardgecoded dann.

Ich glaube man kann das ganze System etwas angenehmer lösen:
UpdatePublisher speichert irgendnen `OrderedDictionary<Type, OrderedDictionary<object, events>>`.

Beim Subscriben muss das eigene Objekt und die Events eingegeben werden.
Zusätzlich kann man spezifizieren, dass die eigenen Events VOR oder NACH den Events von einem beliebigen anderen Datentyp aufgerufen werden (deshalb OrderedDictionary).

In Dispose() werden einfach alle Events vom eigenen Objekt automatisch gelöscht.

Werde das morgen implementieren, dann wirds vielleicht klarer.

funketh commented:
Die GameObjectDatas sollten definitiv nicht von MonoBehaviour erben (um das vermeiden gab es vorher ja extra die komplizierte Generics Datenstruktur).
Man kann aber jedem GameObjectData Objekt automatisch ein simples LinkedBehaviour zuweisen, welches nur ein Attribut hat, das auf das GameObjectData Objekt referenziert.

rrickfox commented:
> Man kann aber jedem GameObjectData Objekt automatisch ein simples LinkedBehaviour zuweisen, welches nur ein Attribut hat, das auf das GameObjectData Objekt referenziert.

Klingt gut, dann bleibt das Feature erhalten.

funketh committed:
use dependency graph instead of weak references

- remove WeakVoidPublisher.cs
- A Type `T` can now specify other types as dependencies
    - when updates are published to instances of `T`, they will wait until all updates of all instances of `T`'s dependencies are done
- unsubscribing is done automatically in GameObjectData.Dispose()

funketh committed:
remove redundant interfaces: IVertex, IGameObjectData

funketh committed:
add LinkedBehaviour again

- this allows you to access GameObjectData from a GameObject
- BUT: LinkedBehaviour is no longer generic. Therefore you may have to cast LinkedBehaviour.data to the appropriate type

funketh committed:
add more comments to Event related code

funketh commented:
> > Man kann aber jedem GameObjectData Objekt automatisch ein simples LinkedBehaviour zuweisen, welches nur ein Attribut hat, das auf das GameObjectData Objekt referenziert.
> 
> Klingt gut, dann bleibt das Feature erhalten.

Man muss jetzt beachten, dass man `LinkedBehaviour.data` manuell casten muss

Shafma commented on Assets/Scripts/Events/UpdatePublisher.cs:
```diff
@@ -0,0 +1,34 @@
+    // Manages the publishing of FixedUpdates
+    public class UpdatePublisher : MonoBehaviour
+    {
+        private static Dictionary<Type, TypePublisher> _publishers { get; } = new Dictionary<Type, TypePublisher>();
```
Nutzen des Dictionaries außer für die Exception?

funketh replied:
Gibt keinen. Werde es zu nem HashSet ändern.

funketh committed:
use HashSet<TypePublisher> instead of a dictionary in UpdatePublisher

- the extra information about the type was not needed

funketh commented:
@rrickfox kannst du es mal reviewen und mir sagen wenns okay ist? Würde dann für dich approven, weil du ja den PR erstellt hast

funketh committed:
add default name generation again
